### ISSUE TITLE
Allow remapping touch controls(mainly right thumbstick)

### ISSUE THREAD

**Issue by andreyorst** [hrydgard/ppsspp#9859]
First of all thanks for great PSP emulator. I've played Monster Hunter, and other games where camera control is important. And quite often we control `Camera` with `D-Pad`. When I play ppsspp on my PC, I usually map joystick's second stick to `D-pad`, to control camera with my right thumb. But when I play on my Android phone, there is no other option but to use `combo keys` to create second `D-Pad` on the right side of the screen. Another optional `Analog stick` to control `D-Pad` would be wery useful.

This is my configuration for MHFU:
![image](https://user-images.githubusercontent.com/19470159/28529373-0e0a8b8c-7099-11e7-85a7-6295e9af2c82.png)
Buttons `1` and `2` mapped to `right` and `left` on `D-pad`  and it is working workaround, but if I'll add two more buttons to control up and down behaviour it will become less convenient, and diagonal control will be still imposible.

**Comment by hrydgard** [comment]
Yeah, that's a good idea. Noted as a feature request and for 1.6, maybe I'll get it into 1.5.

**Comment by andreyorst** [comment]
Ah, I didn't knew about it at all. Definitely will check those "PS3 remasters".

As I was thinking about this second stick, I came to this: this stick should have saparate submenu as `combo keys` do. And it should not just control `D-Pad` but also should feature custom mappings because there are some games like Call Of Duty, which used `cross, triangle, square and circle` to control camera, which was really painful. And I believe that games that use `R` and `L` buttons to rotate camere would also get some benefits from this custom mapped stick.

Maybe it is too much, but it will change the way of playing with touch input.

**Comment by unknownbrackets** [comment]
Just noting that the Vita also allows this (though with its physical right stick):

http://www.guidingtech.com/assets/postimages/2013/07/Vita-Right-Stick.png

Arbitrary mapping would be better (I know of games that use L/R and ▲/X), but a menu of options might be faster to configure and simpler.

Some might map both analogs to the left analog, which could get a little tricky too... aside from the dpad mapping problems.

-[Unknown]

**Comment by jajabinx35** [comment]
Hi all,

Firstly apologies for bumping this thread up again...I'm Also a big fan of ppsspp and after reading this thread, I thought that this is actually a really good idea. 

For example, on my phone (oneplus 5), I map the volume up and down as the L and R buttons (with the awesome options ppsspp offers, you can rotate the screen to landscape reverse so that the volume buttons are on the top)

Now, if there was an option to have a second analogue which you can map to the psp face buttons (X O square triangle), it would make fps games become really playable. Also same with monster Hunter and peace Walker.

Just wanted to share my thought...

**Comment by Stddgv11** [comment]
This!  This could be big and turn soo many shooting games that are way too easy because of screen size auto aim or are frustratingly difficult because of aiming with buttons into hugely enjoyable games the way the second stick does for psp games on the Vita.

This could make so many games playable on the ppsspp that are currently off limits due to the camera being connected to the to buttons instead of a stick.

**Comment by LunaMoo** [comment]
Not really, first of all "right analog stick" mapped to face buttons, would work one of the two ways:
 1) reskin of the face buttons to appear the same as analog stick, but it would still work at best like face buttons,
 2) real right analog stick with axis with an option to map it to face buttons, that in theory should work exactly same as the above, but in PPSSPP's reality would actually work worse as it would have the hardcoded 75% deadzone aka would only react to 75% or higher angles and as long as #11226 stands unfixed it would never work as "good" as 1(which still is terrible and have purely psychological effect).

 The only way to make it better is adding real right analog stick and patch each game separately by right analog game specific cwcheat, but those exist only to few games(such patches are possible to make for every game out there, but it's not automatic, it takes time to do so nobody does that). Few games(HD remasters released for ps3) does support right analog stick natively, but they're actually lying and are just remapping digital functionality to analog which means they work pretty much like 1.

**Comment by andreyorst** [comment]
I've played a lot of games via ppsspp with d-pad mapped to right analog stick on my ps3 controller. All I can say is that this is totally a better experience. Even if camera isn't smooth it is far easier to grab a stick and push it to the direction you want with a quick move of your thumb. And psp games are designed with such camera, which doesn't have option to be moved with real analog stick, and because HD remasters use the same approach as i proposed i think it makes a lot of sense to create secondary mappable analog stick.

**Comment by LunaMoo** [comment]
Except ppsspp can map real physical analogs, this is about a texture on the touchscreen - completely different experience and how that texture looks wouldn't affect how you feel it even if real stick feels better for moving camera than face buttons. It really only has sense as a right analog, since games can be patched to use it.

**Comment by andreyorst** [comment]
> how that texture looks wouldn't affect how you feel it

It would. I can grab virtual stick at the centre, and move it to the direction I need. But I can't put my finger in the middle of D-Pad, because it will trigger one of 4th buttons (or even diagonal motion). So to control camera with a D-Pad I need a precision when I press buttons. But with fake analog, I can grab it and them move around, and for example, without letting it go, move it to the centre and this will stop camera from moving. So yes. Hardcoded 75% area is absolutely fine. Even more, it would be **fantastic**.

If game supports second analog via patch, ppsspp can just apply patch. If not fallback to mappable behaviour.

**Comment by hrydgard** [comment]
Yeah the movement of the touch stick is quite different from four touch buttons so there's definitely some benefit. Some games would even benefit from even more specialized controls...

**Comment by JoaoVitorApolinario** [comment]
Great idea, would be a nice improvement

**Comment by andreyorst** [comment]
From #12182:

>Not to be confused with #9859 which this doesn't fulfill despite that issue name as it's a request for mapping other controls for the analog which this doesn't.

I guess, since this issue was closed, there will be no way to map custom buttons to second analog?

**Comment by LunaMoo** [comment]
Henrik likely got confused since the name of this issue was really not something it really asked for, so I corrected all of that.;3

**Comment by hrydgard** [comment]
Whoops!

**Comment by iwestfall7062** [comment]
I've been googling for a way to map touch screen controls on Android as something other than their standard mapping. Specifically, the right analog stick that's available, mapping it to the d-pad or the X, O, Square, Triangle buttons. Is this what this feature request is for?

**Comment by andreyorst** [comment]
yes, that's what I initially requested.

**Comment by XiaomiRedmi9** [comment]
How about this now?

**Comment by n0p0n3** [comment]
I too am trying to remap the virtual controls, but for the japanese digivice ver portable (working on translation). I've tried to remap the square and triangle buttons to L and R to use the accelerometer as a pedometer via the analog stick emulation like the actual devices (in this psp version you push triangle or square to walk since theres no pedometer). I can add custom buttons to handle the remapping of the dpad and X and O buttons, but since I can't map the square and triangle to L and R, I have to set the analog stick emulation to psp action buttons, which works but also activates the X and O and triggers menus, unless there's a way to set a deadzone over half the analog axis I can't find?
